algorithmic-strategies/problem-set-algorithmic-strategies.md

```diff
@@ -0,0 +1,336 @@
+        [96, 56, 143, 123, 96, 83, 82, 117, 82, 52, 66, 126, 122, 90, 128, 149, 149, 116, 55, 145, 111, 145, 63, 137, 85, 148, 70, 125, 57, 72, 82, 124, 107, 137, 127, 139, 125, 94, 111, 108, 117, 142, 121, 103, 140, 54, 82, 55, 104, 50, 98, 82, 50, 64, 105, 112, 70, 113, 91, 72, 108, 112, 132, 141, 85, 138, 94, 117, 137, 104, 54, 135, 105, 95, 135, 150, 143, 144, 143, 132, 52, 90, 100, 119, 59, 147, 125, 55, 101, 90, 72, 90, 58, 131, 123, 76, 109, 130, 138, 77],
+        200
+    ) == 1657
+```
+
+### !end-question
+### !placeholder
+
+```python
+def knapsack(weights, values, max_weight):
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestZeroOneKnapsack(unittest.TestCase):
+    def test_empty_knapsack(self):
+        self.assertEqual(knapsack([], [], 0), 0)
+
+    def test_small_knapsack(self):
+        self.assertEqual(knapsack([10, 20, 30], [60, 100, 120], 50), 220)
+
+    def test_medium_knapsack(self):
+        self.assertEqual(knapsack([99, 60, 70, 68, 50, 65, 48, 31, 52, 29], [104, 105, 138, 135, 82, 73, 103, 52, 57, 136], 75), 188)
+
+    def test_large_knapsack(self):
+        self.assertEqual(knapsack([92, 39, 56, 81, 59, 38, 71, 91, 73, 34, 13, 27, 10, 49, 56, 90, 66, 91, 47, 52, 46, 43, 69, 53, 90, 99, 84, 41, 33, 89, 46, 78, 29, 16, 58, 44, 11, 87, 26, 13, 11, 63, 81, 51, 12, 90, 29, 100, 93, 43, 52, 65, 45, 46, 45, 98, 74, 58, 60, 100, 42, 66, 56, 37, 80, 68, 91, 63, 65, 73, 59, 10, 41, 30, 88, 79, 11, 94, 20, 89, 83, 70, 47, 75, 19, 15, 33, 34, 94, 65, 66, 44, 62, 95, 45, 17, 92, 93, 27, 43], [96, 56, 143, 123, 96, 83, 82, 117, 82, 52, 66, 126, 122, 90, 128, 149, 149, 116, 55, 145, 111, 145, 63, 137, 85, 148, 70, 125, 57, 72, 82, 124, 107, 137, 127, 139, 125, 94, 111, 108, 117, 142, 121, 103, 140, 54, 82, 55, 104, 50, 98, 82, 50, 64, 105, 112, 70, 113, 91, 72, 108, 112, 132, 141, 85, 138, 94, 117, 137, 104, 54, 135, 105, 95, 135, 150, 143, 144, 143, 132, 52, 90, 100, 119, 59, 147, 125, 55, 101, 90, 72, 90, 58, 131, 123, 76, 109, 130, 138, 77], 200), 1657)
+```
+### !end-tests
+### !hint
+
+- We probably need to add an additional parameter to be able to store and pass memoized data.
+- What data type would be appropriate for storing the memoized data?
+- How can we initialize the storage for the memoized data?
+- Where do we need to update our memoized data?
+- Where do we need to pass our memoized data?
+  
+### !end-hint
+
+### !hint
+
+- What data coming into the function call can be used to uniquely mark _this_ function call? The weights? The values? The max weight? Are any of these redundant?
+- If multiple pieces of data are required, how can we use those as keys into our memoized data? Can we use multi-dimensional storage? Is there a way to pack multiple pieces of data into a single piece of data?
```

Comment:
Is the multi-dimensional storage a red herring? (aka, will a student find a solution if they say "yes" to this question?)

If it is, let's delete it!

```suggestion
- If multiple pieces of data are required, how can we use those as keys into our memoized data? Is there a way to pack multiple pieces of data into a single piece of data?
```

Reply:
They could absolutely do with with a multi-dimensional dict, just like the solution in lcs. I presented the tuple-based single-dimension approach since we haven't really shown many cases where tuples are useful, and I expect many student will be surprised that this works. But two-dimensional dict is perfectly fine.